docs/src/tutorials/getting_started/design_patterns_for_larger_models.jl

```diff
@@ -0,0 +1,645 @@
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,       #src
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE    #src
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER         #src
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,  #src
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE  #src
+# SOFTWARE.                                                                      #src
+
+# # Design patterns for larger models
+
+# JuMP makes it easy to build and solve optimization models. However, once you
+# start to construct larger models, and especially ones that interact with
+# external data sources or have customizable sets of variables and constraints
+# based on client choices, you may find that your scripts become unwieldy. This
+# tutorial demonstrates a variety of ways in which you can structure larger JuMP
+# models to improve their readability and maintainability.
+
+# !!! tip
+#     This tutorial is more advanced than the other "Getting started" tutorials.
+#     It's in the "Getting started" section to give you an early preview of how
+#     JuMP makes it easy to structure larger models. However, if you are new to
+#     JuMP you may want to briefly skim the tutorial, and come back to it once
+#     you have written a few JuMP models.
+
+# ## Overview
+
+# This tutorial uses explanation-by-example. We're going to start with a simple
+# [knapsack model](https://en.wikipedia.org/wiki/Knapsack_problem), and then
+# expand it to add various features and structure.
+
+# ## A simple script
+
+# Your first prototype of a JuMP model is probably a script that uses a small
+# set of hard-coded data.
+
+using JuMP, GLPK
+profit = [5, 3, 2, 7, 4]
+weight = [2, 8, 4, 2, 5]
+capacity = 10
+N = 5
+model = Model(GLPK.Optimizer)
+@variable(model, x[1:N], Bin)
+@objective(model, Max, sum(profit[i] * x[i] for i in 1:N))
+@constraint(model, sum(weight[i] * x[i] for i in 1:N) <= capacity)
+optimize!(model)
+value.(x)
+
+# The benefits of this approach are:
+#  * it is quick to code
+#  * it is quick to make changes.
+
+# The downsides include:
+#  * all variables are global (read [Performance tips](https://docs.julialang.org/en/v1/manual/performance-tips/))
+#  * it is easy to introduce errors, e.g., having `profit` and `weight` be
+#    vectors of different lengths, or not match `N`
+#  * the solution, `x[i]`, is hard to interpret without knowing the order in
+#    which we provided the data.
+
+# ## Wrap the model in a fuction
+
+# A good next step is to wrap your model in a function. This is useful for a few
+# reasons:
+#  * it removes global variables
+#  * it encapsulates the JuMP model and forces you to clarify your inputs and
+#    outputs
+#  * we can add some error checking.
+
+function solve_knapsack_1(profit::Vector, weight::Vector, capacity::Real)
+    @assert length(profit) == length(weight)
```

Comment:
I would make this a DimensionMismatch or such because asserts shouldn’t be used for error checking (they are documented that they could be disabled).